javascript/ql/lib/semmle/javascript/frameworks/Spife.qll

```diff
@@ -0,0 +1,412 @@
+ */
+
+import javascript
+import semmle.javascript.frameworks.HTTP
+
+/**
+ * Provides classes for working with [Spife](https://github.com/npm/spife) applications.
+ */
+module Spife {
+  /**
+   * A call to a Spife method that sets up a route.
+   */
+  private class RouteSetup extends API::CallNode, Http::Servers::StandardRouteSetup {
+    TaggedTemplateExpr template;
+
+    RouteSetup() {
+      this.getCalleeNode().asExpr() = template and
+      API::moduleImport(["@npm/spife/routing", "spife/routing"])
+          .asSource()
+          .flowsToExpr(template.getTag())
+    }
+
+    private string getRoutePattern() {
+      // Concatenate the constant parts of the expression
+      result =
+        concat(Expr e, int i |
+          e = template.getTemplate().getElement(i) and exists(e.getStringValue())
+        |
+          e.getStringValue() order by i
+        )
+    }
+
+    private string getARouteLine() {
+      result = this.getRoutePattern().splitAt("\n").regexpReplaceAll(" +", " ").trim()
+    }
+
+    private predicate hasLine(string method, string path, string handlerName) {
+      exists(string line | line = this.getARouteLine() |
+        line.splitAt(" ", 0) = method and
+        line.splitAt(" ", 1) = path and
+        line.splitAt(" ", 2) = handlerName
+      )
+    }
+
+    API::Node getHandlerByName(string name) { result = this.getParameter(0).getMember(name) }
+
+    API::Node getHandlerByRoute(string method, string path) {
+      exists(string handlerName |
+        this.hasLine(method, path, handlerName) and
+        result = this.getHandlerByName(handlerName)
+      )
+    }
+
+    override DataFlow::SourceNode getARouteHandler() {
+      result = this.getHandlerByRoute(_, _).getAValueReachingSink().(DataFlow::FunctionNode)
+      or
+      exists(DataFlow::MethodCallNode validation |
+        validation = this.getHandlerByRoute(_, _).getAValueReachingSink() and
+        result = validation.getArgument(1).getAFunctionValue()
+      )
+    }
+
+    override DataFlow::Node getServer() { none() }
+  }
+
+  /**
+   * A Spife route handler.
+   */
+  abstract class RouteHandler extends Http::Servers::StandardRouteHandler, DataFlow::FunctionNode {
+    /**
+     * Gets the parameter of the route handler that contains the request object.
+     */
+    DataFlow::ParameterNode getRequestParameter() { result = this.getParameter(0) }
+
+    /**
+     * Gets the parameter of the route handler that contains the context object.
+     */
+    DataFlow::ParameterNode getContextParameter() { result = this.getParameter(1) }
+  }
+
+  /**
+   * A standard Spife route handler.
+   */
+  private class StandardRouteHandler extends RouteHandler, DataFlow::FunctionNode {
+    StandardRouteHandler() { any(RouteSetup setup).getARouteHandler() = this }
+  }
+
+  /**
+   * A function that looks like a Spife route handler.
+   *
+   * For example, this could be the function `function(req, res, next){...}`.
```

Comment:
Documentation doesn't match the implementation.   

Your example has 3 parameters, but the charpred looks for exactly two parameters.  
Looks like the qldoc wasn't changed after copy-pasting.